docs/branching-and-tagging.md

```diff
@@ -0,0 +1,218 @@
+table.newRewrite()
+    .rewriteFiles(ImmutableSet.of(small_file_1,small_file_2), ImmutableSet.of(compacted_file))
+    .toBranch(branch)
+    .commit();
+
+```
+
+#### Reading from branches and tags
+Reading from a branch or tag can be done as usual via the Table Scan API, by passing in a branch or tag in the useRef API. When a branch is passed in, the snapshot that's used is the head of the branch. Note that currently reading from a branch and specifying an asOfSnapshotId in the scan is not supported. 
+
+```java
+// Read from the head snapshot of test-branch
+TableScan branchRead = table.newScan().useRef("test-branch");
+
+// Read from the snapshot referenced by audit-tag
+Table tagRead = table.newScan().useRef("audit-tag");
+```
+
+#### Replacing and fast forwarding branches and tags
+
+The snapshots which existing branches and tags point to can be updated via the `replace` APIs. The fast forward operation is similar to git fast-forwarding. Fast forward can be used to advance a target branch to the head of a source branch or a tag when the target branch is an ancestor of the source. For both fast forward and replace, retention properties of the target branch are maintained by default.
+
+```java
+import org.apache.iceberg.Table;
+
+Table table = catalog.loadTable(...)
+String branch = "test-branch";
+
+// Update "test-branch" to point to snapshot 4
+table.manageSnapshots()
+     .replaceBranch(branch, 4)
+     .commit()
+
+table.manageSnapshots()
+     .replaceBranch(branch, 4)
+     .commit()
+
+String tag = "audit-tag";
+// Replace "audit-tag" to point to snapshot 3 and update its retention
+table.manageSnapshots()
+     .replaceBranch(tag, 4)
+     .setMaxRefAgeMs(1000)
+     .commit()
+
+
+```
+
+#### Updating retention properties
+
+Retention properties for branches and tags can be updated as well.
+Use the setMaxRefAgeMs for updating the retention property of the branch or tag itself. Branch snapshot retention properties can be updated via the setMinSnapshotsToKeep and setMaxSnapshotAgeMs APIs. 
+
+```java
+import org.apache.iceberg.Table;
+
+Table table = catalog.loadTable(...)
+
+String branch = "test-branch";
+// Update retention properties for test-branch
+table.manageSnapshots()
+    .setMinSnapshotsToKeep(branch, 10)
+    .setMaxSnapshotAgeMs(branch, 7200000)
+    .setMaxRefAgeMs(branch, 604800000)
+    .commit();
+
+// Update retention properties for test-tag
+table.manageSnapshots()
+    .setMaxRefAgeMs("test-tag", 604800000)
+    .commit();
+```
+
+#### Removing branches and tags
+
+Branches and tags can be removed via the removeBranch API
+
+```java
+import org.apache.iceberg.Table;
+
+Table table = catalog.loadTable(...)
+
+// Remove test-branch
+table.manageSnapshots()
+     .removeBranch("test-branch")
+     .commit()
+
+// Remove test-tag
+table.manageSnapshots()
+     .removeTag("test-tag")
+     .commit()
+```
+
+### Spark Integration
+
+#### Reading
```

Comment:
Pending https://github.com/apache/iceberg/pull/6651/files for adding writes to this